crates/partition-store/src/partition_store.rs

```diff
@@ -288,9 +291,77 @@ impl PartitionStore {
             key_range,
             key_buffer: BytesMut::new(),
             value_buffer: BytesMut::new(),
+            idempotency_table_disabled: false,
         }
     }
 
+    /// Creates a new PartitionStore and checks if the idempotency table is empty,
+    /// disabling it when empty to avoid unnecessary operations.
+    pub(crate) async fn new_with_idempotency_check(
+        rocksdb: Arc<RocksDb>,
+        data_cf_name: CfName,
+        partition_id: PartitionId,
+        key_range: RangeInclusive<PartitionKey>,
+    ) -> Result<Self> {
+        let mut store = Self {
+            rocksdb,
+            partition_id,
+            data_cf_name,
+            key_range,
+            key_buffer: BytesMut::new(),
+            value_buffer: BytesMut::new(),
+            idempotency_table_disabled: false,
+        };
+
+        // Check if the idempotency table is empty
+        let is_empty = store.is_idempotency_table_empty()?;
+        if is_empty {
+            info!(
+                "Idempotency table for partition {} is empty, disabling all idempotency table code paths for better performance",
+                partition_id
+            );
+            store.idempotency_table_disabled = true;
+        }
+
+        Ok(store)
+    }
+
+    /// Checks if the idempotency table is empty for this partition
+    fn is_idempotency_table_empty(&self) -> Result<bool> {
+        let table = self.table_handle(TableKind::Idempotency)?;
+        let mut opts = ReadOptions::default();
+
+        // Get the bytes representation of the KeyKind::Idempotency
+        let key_kind_bytes = KeyKind::Idempotency.as_bytes();
+
+        // Configure the iterator options
+        opts.set_prefix_same_as_start(true);
+        opts.set_async_io(true);
+        opts.set_total_order_seek(false);
```

Comment:
I am not 100% sure that this would work. I think a problem is that the configured prefix extractor is the fixed_prefix with a prefix length of 10 bytes (key kind + partition id). We would have loved to use the NewCappedPrefixTransform, but it's not yet exposed via the Rust bindings. Only seeking to the key kind might give wrong results w/o `set_total_order_seek(true)`. Worth testing it out.